Consul Config Sourcer Plugin

paladin-devops opened the pull request:
This PR adds a new config sourcer plugin to support dynamic configuration from the Consul KV data store.

Example config source settings:
`waypoint config source-set -type=consul -config=address=$CONSUL_HTTP_ADDR -config=token=$CONSUL_HTTP_TOKEN`

Example `waypoint.hcl` snippet using this plugin:
```hcl
  config {
    env = {
      "TEST" = dynamic("consul", {
        key  = "paladin"
      })
    }
  }
```

This plugin was authored by @mkeeler! 🎉

mkeeler committed:
WIP implementation of a Consul configsourcer plugin

paladin-devops committed:
Add Consul deps.

paladin-devops committed:
Add Consul config sourcer docs.

paladin-devops committed:
Changelog.

catsby commented on builtin/consul/config_sourcer.go:
```diff
@@ -0,0 +1,410 @@
+
+func (t *TLSConfig) toAPITLSConfig() *api.TLSConfig {
+	return &api.TLSConfig{
+		Address:            t.ServerName,
```
I assume the default golang value of `""` is acceptable input here?

paladin-devops replied:
According to the Consul API documentation, this field is optional, so `""` should be acceptable.

from `github.com/hashicorp/consul/api/api.go`:
```
	// Address is the optional address of the Consul server. The port, if any
	// will be removed from here and this will be set to the ServerName of the
	// resulting config.
	Address string
```

catsby commented on builtin/consul/config_sourcer.go:
```diff
@@ -0,0 +1,410 @@
+		TokenFile:  conf.TokenFile,
+		Namespace:  conf.Namespace,
+		Partition:  conf.Partition,
+		TLSConfig:  *conf.TLSConfig.toAPITLSConfig(),
```
nit: `toAPITLSConfig()` only seems to be used in a single place, it seems strange that it returns a pointer to a struct to only then be dereferenced because the 1 place that uses it expects a struct and not a pointer. Is there a reason it's designed this way?

paladin-devops replied:
Not particularly, I've updated this to:
```go
		TLSConfig: api.TLSConfig{
			Address:            conf.TLSConfig.ServerName,
			CAFile:             conf.TLSConfig.CAFile,
			CAPath:             conf.TLSConfig.CAPath,
			CAPem:              conf.TLSConfig.CAPem,
			CertFile:           conf.TLSConfig.CertFile,
			CertPEM:            conf.TLSConfig.CertPEM,
			KeyFile:            conf.TLSConfig.KeyFile,
			KeyPEM:             conf.TLSConfig.KeyPEM,
			InsecureSkipVerify: conf.TLSConfig.InsecureHTTPs,
		},
```

catsby commented on builtin/consul/config_sourcer.go:
```diff
@@ -0,0 +1,410 @@
+	}
+}
+
+type ConsulHTTPAuth struct {
```
Does this need to be exported? Do we anticipate using it outside of the package?

catsby commented on builtin/consul/config_sourcer.go:
```diff
@@ -0,0 +1,410 @@
+	conf, ok := config.(*ConfigSourcerConfig)
+	if !ok {
+		// The Waypoint SDK should ensure this never gets hit
+		return fmt.Errorf("Expected *ConfigSourcerConfig as parameter")
```
nit: in general, error strings should not be capitalized

catsby commented on builtin/consul/config_sourcer.go:
```diff
@@ -0,0 +1,410 @@
+}
+
+func (cs *ConfigSourcer) startBlockingQuery(ctx context.Context, logger hclog.Logger, kvReq reqConfig, lastIndex uint64, opts *api.QueryOptions) {
+	go func() {
```
I think we should rename `startBlockingQuery` to something else, and perhaps a more idiomatic golang way of writing this would be to move the `go` goroutine call to the invocation, so instead of `cs.startBlockingQuery(refreshCtx, log, kvReq, meta.LastIndex, opts)` you have `go cs.startBlockingQuery(refreshCtx, log, kvReq, meta.LastIndex, opts)` and then remove the `go func(){}()` here. The reason being is that it's more intuitive to the reader that calling `startBlockingQuery` is supposed to be operating in a goroutine. 

For example, reading these lines:

```
cs.cache[kvReq.cacheKey()] = cacheVal
cs.startBlockingQuery(refreshCtx, log, kvReq, meta.LastIndex, opts)
```

a reader might think that `cs.startBlockingQuery()` in fact blocks here, and we wait until it returns, because after all the method name has `Blocking` in it... when in reality `startBlockingQuery()` doesn't block at all... it launches a goroutine and returns, regardless of whatever might block in that goroutine.

catsby replied:
I do realize that even in Effective Go there is an example of this style where a function or method immediately launches a goroutine, I just feel it's clearer to the reader to do `go functionName()`. If the method didn't have `Blocking` in the name I may not have even brought this up 🤷

briancain replied:
I agree, I think it makes more sense to start a go routine on this function rather than making the whole body of this function wrapped in a `go func()`.

paladin-devops replied:
@catsby I'm updating this for the function to be called using a goroutine, rather than the entire contents of the function being in an anonymous function, called as a goroutine.

Also, I think that the nomenclature might be adding to the confusion. Consul has "[Blocking Queries](https://developer.hashicorp.com/consul/api-docs/features/blocking)", which are "used to wait for a potential change using long polling." I'll add information about this to the godoc for `startBlockingQuery`, which I may rename to `startConsulBlockingQuery`.

catsby replied:
@paladin-devops that's good context to know about Consul queries, thanks

catsby commented on builtin/consul/config_sourcer.go:
```diff
@@ -0,0 +1,410 @@
+			pair, meta, err := cs.client.Get(kvReq.Key, opts)
+
+			// check if we are being stopped
+			select {
```
nit: this `select` should be the first thing in the `for` loop.

catsby commented on builtin/consul/config_sourcer.go:
```diff
@@ -0,0 +1,410 @@
+
+			// KV entry not updated - do nothing
+			if meta != nil && meta.LastIndex == lastIndex {
+				logger.Trace("KV value unchanged")
```
should we sleep here for `retryInterval` duration?

paladin-devops replied:
The select block at the link below, which is at the end of the for loop, should take care of waiting for us:

https://github.com/paladin-devops/waypoint/blob/5b671f4ee5809a42f2a3a68ed0aafeeccacbe8db/builtin/consul/config_sourcer.go#L333-L337

catsby replied:
The `continue` statement immediately below this `logger.Trace()` call shoots us back up to the top of the `for` loop.

paladin-devops replied:
[Updated this to sleep for 5 seconds](https://github.com/hashicorp/waypoint/pull/4045/commits/22a277f59dd452fc5946c331babb08bcaea4f2cd) before hitting the API again!

catsby commented on builtin/consul/config_sourcer.go:
```diff
@@ -0,0 +1,410 @@
+			cs.mu.Unlock()
+
+			// now determine the next wait index
+			if err == nil {
+				lastIndex = meta.LastIndex
+				failures = 0
+			} else {
+				// reset the last index to 0 to do a non-blocking query
+				lastIndex = 0
+				// Set up for exponental backoff
+				failures++
+
+				retry := retryInterval * time.Duration(failures*failures)
+				if retry > maxBackoffTime {
+					retry = maxBackoffTime
+				}
+				logger.Error("KV Get errored", "error", err, "retry", retry.String())
+				select {
+				case <-time.After(retry):
+				case <-ctx.Done():
+					return
+				}
+			}
```
Nit: reduce `if err == nil {...} else {...}` to just `if err != nil {...}`, less mental stack to keep track of when reading the code. Just a suggestion 

```suggestion
			if err != nil {
				// reset the last index to 0 to do a non-blocking query
				lastIndex = 0
				// Set up for exponental backoff
				failures++

				retry := retryInterval * time.Duration(failures*failures)
				if retry > maxBackoffTime {
					retry = maxBackoffTime
				}
				logger.Error("KV Get errored", "error", err, "retry", retry.String())
				select {
				case <-time.After(retry):
				        continue
				case <-ctx.Done():
					return
				}
			}
			lastIndex = meta.LastIndex
			failures = 0
```

catsby commented on builtin/consul/config_sourcer.go:
```diff
@@ -0,0 +1,410 @@
+	}()
+}
+
+func (cs *ConfigSourcer) stop() error {
```
Do we anticipate ever returning an `error` here?

paladin-devops replied:
No, nothing in this function really warrants returning an error. However, in all other config sourcer implementations I've seen, including in tests, the return type for `stop`, which is called by `StopFunc`, which satisfies the config sourcer interface from the SDK, is indeed `error`, and I elected not to stray from the path this time around. 😄

catsby replied:
👍 having actually looked at the sdk and interface now, this makes senes

catsby commented on builtin/consul/config_sourcer.go:
```diff
@@ -0,0 +1,410 @@
+			} else {
+				// reset the last index to 0 to do a non-blocking query
+				lastIndex = 0
+				// Set up for exponental backoff
```
```suggestion
				// Set up for exponential backoff
```

catsby reviewed (commented):
A lot of nit-pick things, most of which are inconsequential but a few I'd like your thoughts on before 👍

paladin-devops committed:
maint: small error updates, function and type renaming.

paladin-devops committed:
refactor: start Consul blocking query in goroutine.

The function to start a Consul blocking query will now be called via a goroutine, rather than the contents of the function being entirely wrapped inside an anonymous function, called via a goroutine.

paladin-devops committed:
chore: simplify error handling for checking cache in Consul plugin.

paladin-devops committed:
chore: remove single-use helper function for TLS config formatting.

paladin-devops commented on builtin/consul/config_sourcer.go:
```diff
@@ -0,0 +1,412 @@
+
+// ConfigSourcerConfig is used to configure where to talk to Consul, and from
+// where the KV data is to be retrieved
+type ConfigSourcerConfig struct {
```
They're not required to be set as configurations here, because the Consul API will attempt to retrieve the configuration from environment variables as well (not unlike Nomad), and will set reasonable defaults if they're not present. 

https://github.com/hashicorp/consul/blob/main/api/api.go#L439-L441

paladin-devops committed:
chore: return error with status code for Consul plugin config check.

briancain reviewed (approved):
Sweet!

paladin-devops committed:
maint: avoid slamming Consul API.

Update Consul config sourcer plugin to sleep for 5 seconds before hitting the Consul KV API again, if there is no value change.

catsby reviewed (approved):
👍 LGTM
